tidb-binlog/pump_client/client.go

```diff
@@ -318,52 +325,59 @@ func (c *PumpsClient) WriteBinlog(binlog *pb.Binlog) error {
 		}
 	}
 
-	log.Info("[pumps client] write binlog to avaliable pumps all failed, will try unavaliable pumps")
+	log.Info("[pumps client] write binlog to avaliable pumps all failed, will try unavailable pumps")
 	pump, err1 := c.backoffWriteBinlog(req, binlog.Tp, binlog.StartTs)
 	if err1 == nil {
+		choosePump = pump
 		return nil
 	}
-	choosePump = pump
 
 	return errors.Errorf("write binlog failed, the last error %v", err)
 }
 
+// Return directly for non p-binlog.
+// Try every online pump for p-binlog.
 func (c *PumpsClient) backoffWriteBinlog(req *pb.WriteBinlogReq, binlogType pb.BinlogType, startTS int64) (pump *PumpStatus, err error) {
 	if binlogType != pb.BinlogType_Prewrite {
 		// never return error for commit/rollback binlog.
 		return nil, nil
 	}
 
-	unAvaliablePumps := make([]*PumpStatus, 0, 3)
+	allPumps := make([]*PumpStatus, 0, 3)
 	c.RLock()
-	for _, pump := range c.Pumps.UnAvaliablePumps {
-		unAvaliablePumps = append(unAvaliablePumps, pump)
+	for _, pump := range c.Pumps.Pumps {
+		allPumps = append(allPumps, pump)
 	}
 	c.RUnlock()
 
 	var resp *pb.WriteBinlogResp
-	// send binlog to unavaliable pumps to retry again.
-	for _, pump := range unAvaliablePumps {
-		if !pump.IsUsable() {
+	// send binlog to unavailable pumps to retry again.
+	for _, pump := range allPumps {
+		if !pump.ShouldBeUsable() {
```

Comment:
`IsUsable` also considered `errNum`, is that still relevant now?

Reply:
every online pump should be writable, we should try our best to write success in backoffcase or the service is down.
and the `errNum` is really not clear to tell the behavior here with concurrently `WriteBinlog` in client.go.
```
// IsUsable returns true if pump is usable.
func (p *PumpStatus) IsUsable() bool {
    if !p.ShouldBeUsable() {
        return false
    }

    if atomic.LoadInt64(&p.ErrNum) > defaultMaxErrNums {
        return false
    }

    return true
}
```

Reply:
How do we decide when to call `IsUsable` and `ShouldBeUsable` in the future?

Reply:
according to the code before, IsUsable() means the pump is usable really( may return false if reach some write fails although the state is online.